ckanext/s3filestore/uploader.py

```diff
@@ -188,19 +195,28 @@ def get_signed_url_to_key(self, key, extra_params={}):
         # check whether the object exists in S3
         metadata = client.head_object(Bucket=self.bucket_name, Key=key)
 
+        # check whether the object is publicly readable
+        acl = client.get_object_acl(Bucket=self.bucket_name, Key=key)
+        is_public_read = any(
+            grant['Grantee']['Type'] == 'Group'
+            and grant['Grantee'].get('URI', '').endswith('AllUsers')
+            for grant in acl['Grants']
+        )
+
         if self.signed_url_cache_enabled:
             redis_conn = connect_to_redis()
             cache_key = _get_cache_key(key)
             cache_url = redis_conn.get(cache_key)
-            if cache_url:
+            # use cache only if the visibility is still correct
+            if cache_url and is_public_read != _is_presigned_url(cache_url):
                 log.debug('Returning cached URL for path %s', key)
                 return cache_url
             else:
                 log.debug('No cache found for %s; generating a new URL', key)
 
         params = {'Bucket': self.bucket_name,
                   'Key': key}
-        if metadata['ContentType'] != 'application/pdf':
+        if not is_public_read and metadata['ContentType'] != 'application/pdf':
```

Comment:
There's no point in adding it to the parameters being used to generate a presigned URL, because we're about to strip off the query string anyway and throw away all the presigning.

(It's actually not a very efficient way to generate the URL, but there [doesn't seem to be direct Boto support for getting the public URL of an object](https://stackoverflow.com/questions/48608570/python-3-boto-3-aws-s3-get-object-url).)